src/Orleans.Core/Transactions/ITransactionalResource.cs

```diff
@@ -9,40 +9,43 @@ namespace Orleans.Transactions
     /// </summary>
     public interface ITransactionalResource : IEquatable<ITransactionalResource>
     {
+   
         /// <summary>
-        /// Perform the prepare phase of the commit protocol. To succeed the resource
-        /// must have all the writes that were part of the transaction and is able
-        /// to persist these writes to persistent storage.
-        /// <param name="transactionId">Id of the transaction to prepare</param>
-        /// <param name="writeVersion">version of state to prepare for write</param>
-        /// <param name="readVersion">version of state to prepare for read</param>
+        /// Handle a remote request that returns an ack response or an exception.
         /// </summary>
-        /// <returns>Whether prepare was performed successfully</returns>
-        /// <remarks>
-        /// The resource cannot abort the transaction after it has returned true from
-        /// Prepare.  However, if it can infer that the transaction will definitely
-        /// be aborted (e.g., because it learns that the transaction depends on another
-        /// transaction which has aborted) then it can proceed to rollback the aborted
-        /// transaction.
-        /// </remarks>
-        Task<bool> Prepare(long transactionId, TransactionalResourceVersion? writeVersion, TransactionalResourceVersion? readVersion);
+        /// <returns>A task that completes when the response is received</returns>
+        Task Handle(IRequestWithResponse call);
 
         /// <summary>
-        /// Notification of a transaction abort.
+        /// Handle a remote request that returns a response or an exception.
         /// </summary>
-        /// <param name="transactionId">Id of the aborted transaction</param>
-        Task Abort(long transactionId);
+        /// <typeparam name="T">the type of the response</typeparam>
+        /// <returns>A task that completes when the response is received</returns>
+        Task<T> Handle<T>(IRequestWithResponse<T> call);
```

Comment:
Why move from well defined calls to generic handlers.  This introduces more runtime risk and maintenance cost related to future documenting, implementing, refactoring and testing.

Reply:
The point is to allow extensionality, i.e. avoid hard-coding a specific facet interface that all facets have to use (which goes against the idea of using facets for a wide range of purposes, and means we cannot have multiple transaction algorithms available at the same time). The current facet transactional interface (Prepare/Commit/Abort) does not exist the same way in the new transaction algorithm, rather the new facet transactional interface will be (Prepare/Commit/Prepared/Abort/Confirm/Cancel/Collect/Ping), some of which are requests with responses and some of which are one-way messages. 

The mechanism introduced here allows facets to define their own "interfaces" (currently the facet system does not permit that), and therefore allows the new transaction algorithm to be injected. This helps maintenance and keeps `Orleans.Core` free from implementation-specifics.

Runtime risk and maintenance cost are mitigated as follows:
 - the exact same information as in the interface definition, including all the same comments, still exist. They are just expressed using classes instead of an interface, and have moved to the file `TransactionalResourceCommunication.cs` inside `Orleans.Runtime/Transactions`
 (note: I just noticed that I accidentally deleted one of the comments, will put it back)
 - all requests and responses are typed, thus type errors are still caught by the compiler